services-custom/iam-policy-builder/src/main/java/software/amazon/awssdk/policybuilder/iam/IamStatement.java

```diff
@@ -0,0 +1,1187 @@
+         * before you decide to use {@code NotPrincipal}.
+         * <p>
+         * This works the same as calling {@link #addNotPrincipal(IamPrincipalType, String)} multiple times with the same
+         * {@link IamPrincipalType}.
+         * <p>
+         * {@snippet :
+         * IamStatement statement =
+         *     IamStatement.builder()
+         *                 .sid("GrantReadBookContent")
+         *                 .effect(IamEffect.DENY)
+         *                  // This statement denies access to everyone except the books service and operators:
+         *                 .addNotPrincipals(IamPrincipalType.AWS,
+         *                                   Arrays.asList("arn:aws:iam::123456789012:role/books-service",
+         *                                                "arn:aws:iam::123456789012:role/books-operator"))
+         *                 .addAction("s3:GetObject")
+         *                 .addResource("arn:aws:s3:us-west-2:123456789012:accesspoint/book-content/object/*")
+         *                 .build();
+         * }
+         * @see <a href="https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_notprincipal.html">
+         * NotPrincipal user guide</a>
+         */
+        Builder addNotPrincipals(IamPrincipalType iamPrincipalType, Collection<String> notPrincipals);
+
+        /**
+         * Append multiple 
+         * <a href="https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_notprincipal.html">{@code
+         * NotPrincipal}s</a> to this statement, denying access to the principal that is specified.
+         * <p>
+         * Very few scenarios require the use of {@code NotPrincipal}. We recommend that you explore other authorization options
+         * before you decide to use {@code NotPrincipal}.
+         * <p>
+         * This works the same as calling {@link #addNotPrincipal(String, String)} multiple times with the same
+         * {@link IamPrincipalType}.
+         * <p>
+         * {@snippet :
+         * IamStatement statement =
+         *     IamStatement.builder()
+         *                 .sid("GrantReadBookContent")
+         *                 .effect(IamEffect.DENY)
+         *                  // This statement denies access to everyone except the books service and operators:
+         *                 .addNotPrincipals("AWS", Arrays.asList("arn:aws:iam::123456789012:role/books-service",
+         *                                                        "arn:aws:iam::123456789012:role/books-operator"))
+         *                 .addAction("s3:GetObject")
+         *                 .addResource("arn:aws:s3:us-west-2:123456789012:accesspoint/book-content/object/*")
+         *                 .build();
+         * }
+         * @see <a href="https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_notprincipal.html">
```

Comment:
IAM will fail. I generally prefer to let the service do the validation since they can change and we can't (unless the customer upgrades their SDK), but I'm open to disagreement on that.

Reply:
yeah if IAM will fail, then dont think we need to throw error. would be good to mention in javadocs though